### PR TITLE
refactor(output): port autofix to core

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR ports both metavariable interpolation and autofix to the core Semgrep engine, deleting (some of) the `pysemgrep` and `osemgrep` code for both.

## Why:
It's kind of gross that this logic existed only in the CLI side in the first place. Furthermore, this led to a disparity between the core and CLI outputs, where AST-based autofix would be applied in the first, but not textual autofix. This is conceptually hazardous.

## How:
This was somewhat involved, due to the testy nature of the dependencies of all of the data needed to make this change happen. The primary forces at work here were due to the facts that AST-based autofix needs to _consume_ the actual tree of the fix (thus, the `Pattern_match.t`), and it needs to be able to _produce_ the fix information for the ATD semgrep output, which needs to be able to associate each match with its replacement text.

This PR introduces a new processor called `Autofix_processor`, whose goal is to instrument a core scan by populating its autofix information. Notably, now `Core_result` has the `matches_with_fixes` field, which is not just `Pattern_match.t`, but tuples of `Pattern_match.t * Textedit.t option`.

Since this PR aims to make autofix a functionality of the core engine, and the `Core_json_output` module must be able to turn a `Core_result.t` into an `Out.core_output`, we need this data to exist somewhere, so it makes sense to me that it should become part of `Core_result.t`

### Why not put it at the `Pattern_match.t` -> `Pattern_match.t` level?
I did not want to add the fix information into the `Pattern_match.t` type. That felt too far forward.

### Why not put it at the `Core_json_output` and `match_to_match` level?
I felt that this would conflate the idea of producing the JSON output and producing the autofixes themselves. This also felt too "interleaved" -- I wanted a horizontal step of producing _all_ of the `Out.core_match`es, and a _separate_ step which produces and applies the autofixes, across each match, as opposed to a vertical step where we produce a single `Out.core_match` and then autofix it.

This also has dreadful consequences because the `Autofix` module is not available from `Core_json_output`. You would have to pass in a functional argument or something, plus the `autofix` flag.

### Why keep the Python code for `autofix.py` and not apply fixes in the core engine?
That was the original plan. If you apply the autofixes in the core engine, however, this will cause the match report to be wrong, as the CLI will report the matches of the edited file. Thus, the autofixes need to occur after that point.

## Test plan:
Automated tests.

I verified that the new tests fail when they are mangled.

### CODE REVIEW

File: src/core/Core_result.ml [hunk 1]
```diff
@@ -49,7 +49,20 @@ type 'a match_result = {
 [@@deriving show]
 
 type t = {
-  matches : Pattern_match.t list;
+  (* Why is this here?
```
Comment (aryx) [inline src/core/Core_result.ml]: Maybe put
(* old: matches: Pattern_match.t list;
 * Why do we added the fixes here?
 *  ....
 *)

Comment (aryx) [reply src/core/Core_result.ml]: BTW great comment.

Comment (aryx) [reply src/core/Core_result.ml]: I would maybe add a few alt: comment, explaining design alternative here,
like the stuff you wrote in the PR description.

File: src/core/Core_result.ml [hunk 2]
```diff
@@ -261,8 +274,13 @@ let make_final_result
     (skipped_rules : Rule.invalid_rule_error list) (scanned : Fpath.t list)
     ~rules_parse_time =
   (* contenating information from the match_result list *)
-  let matches =
-    results |> List.concat_map (fun (x : _ match_result) -> x.matches)
+  let matches_with_nullary_fixes =
+    results
+    |> List.concat_map (fun (x : _ match_result) -> x.matches)
+    (* These fixes are initially None, and will be populated with fixes
```
Comment (aryx) [inline src/core/Core_result.ml:286]: nice comment again

File: src/core/Mini_rule.ml [hunk 3]
```diff
@@ -48,6 +48,7 @@ type rule = {
    *)
   pattern_string : string;
   fix : string option;
+  fix_regexp : (string * int option * string) option;
```
Comment (aryx) [inline src/core/Mini_rule.ml]: maybe introduce an intermediate type because those string * int option * string are a bit mysterious
maybe something in Rule.ml like type fix_regexp = {
   field1:
   field2:
  field3:
  }

File: src/core/Pattern_match.ml [hunk 4]
```diff
@@ -153,6 +153,7 @@ and rule_id = {
    * mini_rule? *)
   message : string;
   fix : string option;
```
Comment (aryx) [inline src/core/Pattern_match.ml:155]: I think we should fix the TODO above at some point (not in this PR though) and remove
all those fields from rule_id

**Review by aryx — CHANGES_REQUESTED** [review]
This looks great!
But please can you move the autofix stuff in a separate tests/autofix/python/
I'd rather have tests/patterns/ just doing the matching thing.

**Comment by brandonspark** [comment]
@aryx should be good to review now (this is blocking my nosemgrep PR too)

**Comment by aryx** [comment]
not deleting autofix.py anymore?
What happened?

**Comment by aryx** [comment]
and why the fix_regex got removed from the snapshots?

**Comment by brandonspark** [comment]
@aryx This PR went through many iterations. In an earlier iteration, we passed an `-autofix` command to core, which then applied the autofix.

Unfortunately, this mangles the CLI, which later reads the file at the specified positions to report the finding. If the autofix happens in core, then it will read the updated file, and thus report a finding post-fix.

The easiest way to make this compatible was to retain the CLI as the part which actually _applies_ the autofix, but to make the core engine have the sole responsibility of generating the autofixes, which this PR does.

File: js/engine/tests/__snapshots__/index.test.js.snap [hunk 5]
```diff
@@ -86,7 +86,7 @@ exports[`yaml parser interpolates metavariables in rule message 1`] = `
           },
           "extra": {
             "engine_kind": "OSS",
-            "message": "Print content is $A",
+            "message": "Print content is "Hello World"",
```
Comment (aryx) [inline js/engine/tests/__snapshots__/index.test.js.snap:89]: hmm does this work? the double quotes should be escaped no?

File: src/engine/Unit_engine.mli [hunk 6]
```diff
@@ -5,34 +5,17 @@
...
-   - language of type Lang.t e.g. Lang.Ruby
-   - subdirectory containing the test cases e.g. "ruby"
-   - language extension of the target files e.g. ".rb"
-
-   Each language folder contains pairs of files:
-   - foo.rb: target file for the test case "foo".
-   - foo.sgrep: target file for the test case "foo". If missing, it must
-     exist in the polyglot folder.
-*)
-val make_lang_regression_tests :
```
Comment (aryx) [inline src/engine/Unit_engine.mli:34]: I think this is used in semgrep-pro, so I don't think you can remove it from the .mli

**Review by aryx — APPROVED** [review]
Very nice!
Quite challenging in the end ...
Maybe @nmote has some feedback given he wrote lots of this code.

File: js/engine/tests/__snapshots__/index.test.js.snap (hunk 5 above)
Comment (brandonspark) [reply js/engine/tests/__snapshots__/index.test.js.snap:89]: These are generated, so I'm not sure, but I think it's OK.

Comment (aryx) [reply js/engine/tests/__snapshots__/index.test.js.snap:89]: it's not ok, this is not valid json, so whatever generate those, if it's us, it's not good.

Comment (brandonspark) [reply js/engine/tests/__snapshots__/index.test.js.snap:89]: The one thing I don't understand is, as far as I can see, this is generated directly by the Jest library, generating the JSON from actual Javascript objects. This wasn't produced by us. So I can't see why this would have been our fault (it seems like it would be an issue with Jest if anything)

Comment (brandonspark) [reply js/engine/tests/__snapshots__/index.test.js.snap:89]: Seems this is intended behavior from Jest: https://github.com/jestjs/jest/issues/13776

We could change this, but I don't care altogether too much if it isn't our fault.

Comment (aryx) [reply js/engine/tests/__snapshots__/index.test.js.snap:89]: ah ok. thanks for looking. We're good to go then.

File: src/engine/Match_tainting_mode.ml [hunk 7]
```diff
@@ -104,15 +104,6 @@ module DataflowY = Dataflow_core.Make (struct
   let short_string_of_node n = Display_IL.short_string_of_node_kind n.F2.n
 end)
 
-let convert_rule_id (id, _tok) =
```
Comment (aryx) [inline src/engine/Match_tainting_mode.ml:107]: @brandonspark why did you remove this code?
@IagoAbal do you remember why we had this code in the first place?

**Review by aryx — APPROVED** [review]
impressive

**Comment by aryx** [comment]
I'll fix semgrep-pro

File: src/osemgrep/cli_scan/Scan_subcommand.ml [hunk 8]
```diff
@@ -620,6 +623,12 @@ let run_scan_files (conf : Scan_CLI.conf) (profiler : Profiler.t)
           (String_utils.unit_str (List.length cli_output.paths.scanned) "file")
           (String_utils.unit_str (List.length cli_output.results) "finding"));
 
+    (* step 6: apply autofixes *)
+    (* this must happen posterior to reporting matches, or will report the
+       already-fixed file
+    *)
+    Autofix.apply_fixes_of_core_matches res.core.results;
```
Comment (aryx) [inline src/osemgrep/cli_scan/Scan_subcommand.ml:630]: forgot the if conf.autofix here.
Fixed in https://github.com/semgrep/semgrep/pull/9318
